apps/rig_inbound_gateway/lib/rig_inbound_gateway_web/event_buffer.ex

```diff
@@ -0,0 +1,135 @@
+defmodule RigInboundGatewayWeb.EventBuffer do
+  @moduledoc """
+  Buffers events up to a configurable capacity.
+
+  As soon as the buffer's capacity is reached, newly added items overwrite the oldest ones.
+  """
+  alias __MODULE__
+  alias RigCloudEvents.CloudEvent
+
+  defstruct [:capacity, :events, :write_pointer]
+
+  @doc """
+  creates a new EventBuffer with the given capacity
+  """
+  @spec new(number) :: EventBuffer.t()
+  def new(capacity) do
+    %EventBuffer{capacity: capacity, events: [], write_pointer: 0}
+  end
+
+  @doc """
+  provides the capacity of a given EventBuffer
+  """
+  @spec capacity(EventBuffer.t()) :: number
+  def capacity(buffer) do
+    buffer.capacity
+  end
+
+  @doc """
+  provides all events of a given EventBuffer
+  """
+  @spec all_events(EventBuffer.t()) :: List.t()
+  def all_events(buffer) do
+    buffer.events
+  end
+
+  @doc """
+  writes a value to the given EventBuffer, overwrites the oldest value if EventBuffer is full
+  """
+  @spec add_event(EventBuffer.t(), CloudEvent.t()) :: EventBuffer.t()
+  def add_event(
+        %{
+          write_pointer: write_pointer,
+          capacity: capacity,
+          events: events
+        } = event_buffer,
+        event
+      ) do
+    events = insert_or_update(events, write_pointer, event)
+    write_pointer = rem(write_pointer + 1, capacity)
+
+    %EventBuffer{
+      event_buffer
+      | write_pointer: write_pointer,
+        events: events
+    }
+  end
+
+  def events_since(
+        %{
+          capacity: capacity,
+          write_pointer: write_pointer,
+          events: events
+        } = _event_buffer,
+        event_id
+      ) do
+    case event_id do
+      "first_event" ->
+        new_events = get_new_events(events, 0, write_pointer, capacity)
+
+        case new_events do
+          [] ->
+            {:ok, [events: [], last_event_id: event_id]}
+
+          _ ->
+            {:ok, [events: new_events, last_event_id: CloudEvent.id!(Enum.at(new_events, -1))]}
+        end
+
+      _ ->
+        case Enum.find_index(events, fn x -> CloudEvent.id!(x) == event_id end) do
+          nil ->
+            next_event = Enum.at(events, 0)
+            {:no_such_event, [not_found_id: event_id, last_event_id: CloudEvent.id!(next_event)]}
+
+          event_index ->
+            read_pointer = rem(event_index + 1, capacity)
+            new_events = get_new_events(events, read_pointer, write_pointer, capacity)
+
+            case new_events do
```

Comment:
Perhaps this (and also the same `case` above) could be extracted to function with pattern matching and thus avoid deeply nested `case`s.